test/integration/logic/RuleTest.java

```diff
@@ -381,4 +385,56 @@ public void rule_has_explicit_materialises_when_missing() throws IOException {
             }
         }
     }
+
+    @Test
+    public void rule_that_cannot_be_satisfied_throws_an_error() throws IOException {
+        Util.resetDirectory(directory);
+
+        try (RocksGrakn grakn = RocksGrakn.open(directory)) {
+            grakn.databases().create(database);
+            try (RocksSession session = grakn.session(database, Arguments.Session.Type.SCHEMA)) {
+                try (RocksTransaction txn = session.transaction(Arguments.Transaction.Type.WRITE)) {
+                    final ConceptManager conceptMgr = txn.concepts();
+
+                    final EntityType person = conceptMgr.putEntityType("person");
+                    final EntityType dog = conceptMgr.putEntityType("dog");
+                    final AttributeType name = conceptMgr.putAttributeType("name", AttributeType.ValueType.STRING);
+                    person.setOwns(name);
+
+                    assertThrowsWithMessage(() -> txn.logic().putRule(
+                            "dogs-are-named-fido",
+                            Graql.parsePattern("{$x isa dog;}").asConjunction(),
+                            Graql.parseVariable("$x has name 'fido'").asThing()),
+                                            GraknException.of(RULE_CANNOT_BE_SATISFIED, "dogs-are-named-fido", "$x").getMessage());
+                }
+            }
+        }
+    }
+
+    @Test
+    public void rule_that_cannot_be_inserted_throws_an_error() throws IOException {
+        Util.resetDirectory(directory);
+
+        try (RocksGrakn grakn = RocksGrakn.open(directory)) {
+            grakn.databases().create(database);
+            try (RocksSession session = grakn.session(database, Arguments.Session.Type.SCHEMA)) {
+                try (RocksTransaction txn = session.transaction(Arguments.Transaction.Type.WRITE)) {
+                    final ConceptManager conceptMgr = txn.concepts();
+
+                    final EntityType animal = conceptMgr.putEntityType("animal");
+                    final EntityType person = conceptMgr.putEntityType("person");
+                    final EntityType dog = conceptMgr.putEntityType("dog");
+                    final AttributeType name = conceptMgr.putAttributeType("name", AttributeType.ValueType.STRING);
+                    person.setOwns(name);
+                    person.setSupertype(animal);
+                    dog.setSupertype(animal);
+
+                    assertThrows(() -> txn.logic().putRule(
+                            "animals-are-named-fido",
+                            Graql.parsePattern("{$x isa animal;}").asConjunction(),
+                            Graql.parseVariable("$x has name 'fido'").asThing()));
+                }
+            }
+        }
+    }
```

Comment:
can we also add a 1-4 tests for `insert` in `TypeResolver` please? Maybe if you can think of some slightly non-obvious cases and a few obvious ones

Reply:
Added rules to test this.